topology translation of in-tree vSphere volume to vSphere CSI

divyenpatel opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

Optionally add one or more of the following kinds if applicable:
/kind regression

#### What this PR does / why we need it:
vSphere CSI driver does not understand legacy topology labels in-tree vSphere plugin is using.
This PR is adding topology translation from in-tree vSphere volume to vSphere CSI.

This PR is translating topology GA/topology beta labels to `topology.csi.vmware.com/region` and `topology.csi.vmware.com/zone`.



#### Special notes for your reviewer:
Pre-req for topology translation support.
vSphere CSI Driver must be installed using following vSphere Config Secret

```ini
[Labels]
region = k8s-region
zone = k8s-zone
[TopologyCategory "k8s-region"]
Label = "topology.csi.vmware.com/region"
[TopologyCategory "k8s-zone"]
Label = "topology.csi.vmware.com/zone"
```

Unit test execution

```sh
% go test -v -run  TestTranslatevSphereInTreeStorageClassToCSI
=== RUN   TestTranslatevSphereInTreeStorageClassToCSI
    vsphere_volume_test.go:113: Testing expect error when sc is nil
    vsphere_volume_test.go:113: Testing translate with no parameter
    vsphere_volume_test.go:113: Testing translate with unknown parameter
    vsphere_volume_test.go:113: Testing translate with storagepolicyname and datastore
    vsphere_volume_test.go:113: Testing translate with fstype
    vsphere_volume_test.go:113: Testing translate with storagepolicyname and fstype
    vsphere_volume_test.go:113: Testing translate with no parameter and allowedTopology
    vsphere_volume_test.go:113: Testing translate with storagepolicyname and allowedTopology
    vsphere_volume_test.go:113: Testing translate with storagepolicyname and allowedTopology beta labels
    vsphere_volume_test.go:113: Testing translate with raw vSAN policy parameters, datastore and diskformat
    vsphere_volume_test.go:113: Testing translate with all parameters
--- PASS: TestTranslatevSphereInTreeStorageClassToCSI (0.00s)
PASS
ok      k8s.io/csi-translation-lib/plugins      0.381s
```

```sh
 % go test -v -run  TestTranslateVSphereInTreePVToCSI
=== RUN   TestTranslateVSphereInTreePVToCSI
    vsphere_volume_test.go:464: Testing expect error when in-tree vsphere PV is nil
    vsphere_volume_test.go:464: Testing translate valid vSphere in-tree PV to vSphere CSI PV
    vsphere_volume_test.go:464: Testing translate valid vSphere in-tree PV with beta topology labels to vSphere CSI PV
    vsphere_volume_test.go:464: Testing translate valid vSphere in-tree PV with GA topology labels to vSphere CSI PV
--- PASS: TestTranslateVSphereInTreePVToCSI (0.00s)
PASS
ok      k8s.io/csi-translation-lib/plugins      0.178s
```

```sh
% go test -v -run  TestTranslateVSphereCSIPVToInTree
=== RUN   TestTranslateVSphereCSIPVToInTree
    vsphere_volume_test.go:276: Testing expect error when pv is nil
    vsphere_volume_test.go:276: Testing expect error when pv.Spec.CSI is nil
    vsphere_volume_test.go:276: Testing translate valid vSphere CSI PV to vSphere in-tree PV
    vsphere_volume_test.go:276: Testing translate valid vSphere CSI PV with topology Node Affinity rules to vSphere CSI PV with topology labels
--- PASS: TestTranslateVSphereCSIPVToInTree (0.00s)
PASS
ok      k8s.io/csi-translation-lib/plugins      0.365s
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
topology translation of in-tree vSphere volume to vSphere CSI.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

divyenpatel committed:
topology translation for vSphere CSI

shalini-b commented on staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go:
```diff
@@ -206,3 +227,109 @@ func (t *vSphereCSITranslator) GetCSIPluginName() string {
+				}
+			} else {
+				// Other topologies are passed through unchanged.
+				newExp = exp
```
We did not support any other topologies in VCP right?

divyenpatel replied:
yes used common `translateAllowedTopologies` function, all other plugin is using.

shalini-b commented on staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go:
```diff
@@ -206,3 +227,109 @@ func (t *vSphereCSITranslator) GetCSIPluginName() string {
+					Key:    zonekey,
+					Values: exp.Values,
+				}
+			} else if exp.Key == v1.LabelFailureDomainBetaRegion || exp.Key == v1.LabelTopologyRegion {
```
SC Allowed topologies did not allow region labels right? Why is this needed?

divyenpatel replied:
Yes, and the vSphere in-tree plugin does not read region while creating volume.
https://github.com/kubernetes/kubernetes/blob/06e107081e4e0182efb9daa0daf7f2ae12272704/pkg/volume/vsphere_volume/vsphere_volume.go#L370-L376

I am removing the translation of region then.

shalini-b commented on staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:
```diff
@@ -48,6 +48,16 @@ func TestTranslatevSphereInTreeStorageClassToCSI(t *testing.T) {
+			Values: []string{"zone-a"},
+		},
+		{
+			Key:    vSphereCSITopologyRegionKey,
```
Region label wasn't allowed in SC allowed topologies right?

divyenpatel replied:
removing it

shalini-b commented on staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:
```diff
@@ -194,6 +204,84 @@ func TestTranslateVSphereCSIPVToInTree(t *testing.T) {
+						},
+					},
+					AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+					NodeAffinity: &v1.VolumeNodeAffinity{
```
The In-tree PV will have node affinity terms too?

divyenpatel replied:
I am not sure about it, but if at all we have node affinity rules present on the PV, we need to translate it so the schedular can use the node with topology keys applied by the vSphere CSI Driver.

shalini-b commented on staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:
```diff
@@ -264,6 +352,65 @@ func TestTranslateVSphereInTreePVToCSI(t *testing.T) {
+			intreePV: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "pvc-d8b4475f-2c47-486e-9b57-43ae006f9b59",
+					Labels: map[string]string{"failure-domain.beta.kubernetes.io/zone": "z1", "failure-domain.beta.kubernetes.io/region": "r1"},
```
Should we have another test for In-tree PV with GA labels?

divyenpatel replied:
yes

divyenpatel commented:
@shalini-b I have addressed your review comments here - https://github.com/kubernetes/kubernetes/pull/108611/commits/7aa43796c2ebfd58a35a6f4a8cb925cbb8c29740

shalini-b commented:
/approve

xing-yang commented:
/lgtm

xing-yang commented:
/assign @msau42 @Jiawei0227

xing-yang commented:
/assign @gnufied

msau42 commented:
/lgtm
/approve
